### PR TITLE
feat: test mode

### PR DESCRIPTION
**Pull request by hay-kot**

### CODE REVIEW

File: tests/cli.test.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash 
+export SCAFFOLD_NO_CLOBBER="true" 
+export SCAFFOLD_OUT="gen" 
+export SCAFFOLD_DIR=".scaffold,.examples" 
+
+go run main.go new --test cli
```
Comment (alexeagle) [inline tests/cli.test.sh:6]: I think we'll need more than one test case. Let's say this example had a couple of "features". Then I would need to run `new` to scaffold out several entries in the matrix like `{featureA, featureB, featureA+B}` to have test coverage for mistakes in the scaffold that can only be observed when those features interact in the generated app, as well as verify each feature works in isolation.

So maybe the yaml provides named test cases? Then I would reference the names here like  `--test=case1`?

**Comment by hay-kot** [comment]
@alexeagle, took another stab with a different API. 

- Instead of the `test` key we'd now use a `tests` key
- Different command for tests instead of using a flag, so `scaffold test ...`
- Allow for rendering into an in memory file system
- Allow for outputting an AST (this is how we do our internal testing)
- Allow for running specific cases with `--case` flag

Example AST Output

```bash
scaffold test --log-level="panic" --case="default" --memfs --ast <scaffold-name>
scaffold-test-3811:  (type=dir)
        main.go:  (type=file)
                package main

                import (
                        "fmt"
                )

                func main() {
                        fmt.Println("Hello, World!")
                }
```

**Comment by alexeagle** [comment]
Neat! Thanks :)

An idea for generalizing just a bit:

Since the `tests` block appears in the same scaffold file that users might inspect to see "what prompts does this ask", do you think this feature might also be useful for users to discover the tests, but not use them to assert correctness. Instead the user could skip some of the prompts by choosing "canned" answers for some or all of them. In that case just renaming `tests` to `presets` would make this appear more useful for that use case. 

Personally I wouldn't add a new `test` command here, because you're not giving a way for me to exercise the system-under-test nor make assertions. In the "Arrange, Act, Assert" model, this only provides "Arrange". I think the `new` command with a flag would keep complexity lower, with proposal above

`scaffold new https://github.com/hay-kot/scaffold-go-cli --preset my-preset-1`

**Review by alexeagle — APPROVED** [review]
Neat! excited to try this out :)

**Comment by hay-kot** [comment]
I'm pretty happy with how the API turned out. Going to write some better documentation and write a few more tests this weekend so hopefully it'll be merged and released by Monday.

Thanks for all your feedback, it was _very_ helpful!

**Pull request merged**